File: object_oriented_programming/C#/comp/libcomp.cs
using System;
using System.Collections.Generic;

namespace libcomp{

    public abstract class Device{
        protected string _Name;

        public Device(string name){
            _Name=name;
        }
        public Device(){
            
        }

        public abstract void StartComputer(string ip);
        public abstract void ShutDownComputer(string ip);
    }
    public class Computer : Device
{
    private string _IpAddress;
    private string _BiosName; 
    private static int _counter=0;

    public Computer(string ipAddress, string biosName, string name): base(name)
    {
        _IpAddress = ipAddress;
        _BiosName = biosName;
        
        _counter ++;
    }
    
    public Computer(string biosName, string name): base(name){  //constructor without ip
        _BiosName= biosName;
    }
    
    public Computer(): base(){  // default constructor   commented because error with abstract class
        _counter++;
    }
    
    public override void StartComputer(string ip){  // computer turns on, ip visible
        _IpAddress=ip;
    }
    
    public override void ShutDownComputer(string ip){  // computer turns off, ip not visible
        _IpAddress= null;
    }
    
    public static int getcompsNum(){ // shows number of computers
        return _counter;
    }
    
    public string IpAddress{
        get {return this._IpAddress;}
        set {this._IpAddress= value;}
    }
    
    public string BiosName{
        get {return this._BiosName;}
        set {this._BiosName= value;}
    }
    
    public string Name{
        get {return this._Name;}
        set {this._Name= value;}
    }
}          //end of class computer

public class Server : Computer{
    
    private string _type;
    
    public Server(string ipAddress, string biosName, string name, string type) : base(ipAddress, biosName, name){
        this.Type = type;                   
    }
    
    public Server(){
    }
    
    public string Type{
        get { return _type;}
        set { _type = value; }
    }
    
    public override void StartComputer(string ip){
        IpAddress= ip;
    }
    
    public override void ShutDownComputer(string ip){
        Console.WriteLine("Are you sure you want to shut down the server? (y/n)");
        string sure= Console.ReadLine();
        if(sure == "y" || sure == "n"){
            if(sure =="y"){
                IpAddress=null;
            }else{
                Console.WriteLine("Server still on");
            }
        }else{
            Console.WriteLine("invalid option");
        }
    }
}         //end of class server


}

File: object_oriented_programming/C++/object_oriented_programming/libcomp.cpp
#include "libcomp.h"
#include <iostream>

int Computer::counter = 0;

Computer::Computer(std::string ip, std::string bios, std::string name) 
    : ipAddress(ip), biosName(bios), name(name) {
    counter++;
}

Computer::Computer(std::string bios, std::string name) 
    : biosName(bios), name(name) {
    counter++;
}

Computer::Computer() {
    counter++;
}

void Computer::startComputer(const std::string& ip) {
    ipAddress = ip;
}

void Computer::shutDownComputer() {
    ipAddress = "";
}

std::string Computer::getIpAddress() const { return ipAddress; }
void Computer::setIpAddress(const std::string& ip) { ipAddress = ip; }

std::string Computer::getBiosName() const { return biosName; }
void Computer::setBiosName(const std::string& bios) { biosName = bios; }

std::string Computer::getName() const { return name; }
void Computer::setName(const std::string& name) { this->name = name; }

int Computer::getCompsNum() { return counter; }

Server::Server(std::string ip, std::string bios, std::string name, std::string type) 
    : Computer(ip, bios, name), type(type) {}

Server::Server() {}

std::string Server::getType() const { return type; }
void Server::setType(const std::string& type) { this->type = type; }

void Server::shutDownComputer() {
    std::cout << "Are you sure you want to shut down the server? (y/n): ";
    char choice;
    std::cin >> choice;
    if (choice == 'y') {
        ipAddress = "";
    } else {
        std::cout << "Server still on\n";
    }
}
